### PR TITLE
Adding Office-Runtime and Custom-Functions-Mapping

### PR DESCRIPTION
**Pull request by mscharlock**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/office-js/index.d.ts [hunk 1]
```diff
@@ -17198,9 +17198,109 @@ declare namespace OfficeExtension {
...
+       *
+       * [Api set: Dialog 1.X]
+       */
+      onClose?: () => void;
+  
+      /**
+       *
+       * Callback that is run when the dialog sends an error
+       *
+       * [Api set: Dialog 1.X]
```
Comment (Zlatkovsky) [inline types/office-js/index.d.ts]: For now we should remove the API set entirely and just mark it (here and elsewhere) as beta

File: types/office-runtime/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Typescript Version: 2.4
+
+/*
+office-runtime
+Copyright (c) Microsoft Corporation
+*/
+
+declare namespace OfficeRuntime {
+    function displayWebDialog(url: string, options: DisplayWebDialogOptions): Promise<Dialog>;
```
Comment (Zlatkovsky) [inline types/office-runtime/index.d.ts]: These two should have comments

File: types/office-js/index.d.ts [hunk 3]
```diff
@@ -17198,9 +17198,109 @@ declare namespace OfficeExtension {
 ////////////////////////////////////////////////////////////////
 
 
+
 ////////////////////////////////////////////////////////////////
 
 
+
+////////////////////////////////////////////////////////////////
+//////////////// Begin OfficeRuntime namespace /////////////////
```
[the remark this answers is not in the document]
Comment (Zlatkovsky) [reply types/office-js/index.d.ts]: Actually, I now see you have it separately later as well.  There is no need for duplication, remove it from here and just leave it there (but the comments here probably apply to the other one too)

File: types/office-runtime/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+       *
+       * Provides options for how a dialog is displayed for Excel Custom Functions only. 
+       *
+       */
+    interface DisplayWebDialogOptions {
+      /**
+       * @beta
+       * Determines whether the dialog box displays as a popup (false) or within an IFrame (true). This setting is only applicable to custom functions running on Excel Online. 
+       *
+       * [Api set: Dialog 1.2]
```
Comment (Zlatkovsky) [inline types/office-runtime/index.d.ts]: Remove API sets

File: types/office-runtime/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+       /**
+       * @beta
+       * Method that closes a dialog box. 
+       */
+      close(): Promise<void>;
+    }
+  
+       /**
+       *
+       * Provides options for how a dialog is displayed for Excel Custom Functions only.
```
Comment (Zlatkovsky) [inline types/office-runtime/index.d.ts]: It's not for custom functions only

File: types/office-runtime/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+       * Callback that is run when the dialog box is closed.
+       *
+       * 
+       */
+      onClose?: () => void;
+  
+      /**
+       * @beta
+       * Callback that is run when the dialog box sends an error.
+       *
```
Comment (Zlatkovsky) [inline types/office-runtime/index.d.ts]: Nit, a few too many *s...

File: types/office-runtime/office-runtime-tests.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+OfficeRuntime.AsyncStorage.setItem("foo", "bar").then(error => console.log(error)) 
+
+OfficeRuntime.AsyncStorage.removeItem("foo"); 
+OfficeRuntime.AsyncStorage.removeItem("foo", () => {
+   //do a thing
+})
+
+OfficeRuntime.AsyncStorage.clear();
+OfficeRuntime.AsyncStorage.clear(() => {
+    //do a thing
```
Comment (Zlatkovsky) [inline types/office-runtime/office-runtime-tests.ts]: Probably best to replace `do a thing` with something slightly less colloquial :-)

File: types/office-runtime/office-runtime-tests.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+OfficeRuntime.displayWebDialog("https://localhost:3000", { 
+    height: '50', 
+    width: '50%',
+    displayInIFrame: false,
+    hideTitle: true, 
+        onMessage: (message, dialog) => { 
+            console.log(message); 
+             dialog.close(); 
+             }
```
Comment (Zlatkovsky) [inline types/office-runtime/office-runtime-tests.ts]: nit, formatting

File: types/office-runtime/tslint.json [hunk 9]
```diff
@@ -0,0 +1,79 @@
+{
+    "extends": "dtslint/dt.json", 
+    "rules": {
+        "adjacent-overload-signatures": false,
+        "array-type": false,
+        "arrow-return-shorthand": false,
+        "ban-types": false,
+        "callable-types": false,
```
Comment (Zlatkovsky) [inline types/office-runtime/tslint.json]: again, that seems like too many rules...

File: types/office-runtime/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+       */
+      width?: string;
+  
+      /**
+       * @beta
+       * True if title is hidden from the dialog box. 
+       *
+       * [Api set: Dialog 1.2]
+       */
+      hideTitle?: boolean;
```
Comment (Zlatkovsky) [inline types/office-runtime/index.d.ts]: It turns out that we should remove this one (`hideTitle`).  We will NOT be supporting it for now.

File: types/custom-function-mapping/tslint.json [hunk 11]
```diff
@@ -0,0 +1,79 @@
+{
+    "extends": "dtslint/dt.json", 
+    "rules": {
```
[the remark this answers is not in the document]
Comment (mscharlock) [reply types/custom-function-mapping/tslint.json]: I'm modelling off of previous tslints from Office but I agree it's a lot of things. Though, having looked through some other types which are similarly sized to ours, they also use all of this?

File: types/custom-function-runtime/custom-function-runtime-tests.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
+function ADD10(n: number) {
+    return n + 10;
+}
+
+declare const CustomFunctionMappings: { 
+    ["addTen"]: ADD10();
```
Comment (Zlatkovsky) [inline types/custom-function-runtime/custom-function-runtime-tests.ts]: You don't need the `[` `]` around the "addTen"

File: types/custom-function-runtime/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        setResult: (value: T) => void;
+    }
+
+    interface CancelableHandler {
+        /*
+        * Handles what should occur when a custom function is canceled. 
+        * @beta 
+        */
+        onCanceled: () => void;
+        }
```
Comment (Zlatkovsky) [inline types/custom-function-runtime/index.d.ts]: Nit: indentation

File: types/office-js/index.d.ts [hunk 14]
```diff
@@ -17198,9 +17198,12 @@ declare namespace OfficeExtension {
 ////////////////////////////////////////////////////////////////
 
 
+
```
Comment (Zlatkovsky) [inline types/office-js/index.d.ts]: You should just revert this file

File: types/custom-function-runtime/tslint.json [hunk 15]
```diff
@@ -0,0 +1,79 @@
+{
+    "extends": "dtslint/dt.json", 
+    "rules": {
+        "adjacent-overload-signatures": false,
```
Comment (Zlatkovsky) [inline types/custom-function-runtime/tslint.json]: Still seems like a lot of lint rules... though I suppose it's not necessarily harmful, just unnecessary

File: types/office-runtime/index.d.ts [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    */
+    function displayWebDialog(url: string, options: DisplayWebDialogOptions): Promise<Dialog>;
+    
+    /*
+    * Defines the AsyncStorage constant which allows asynchronous, global, and persistent key-value storage which can be used by Excel custom functions. 
+    */
+    
+    const AsyncStorage: AsyncStorage;
+  
+    // tslint:disable-next-line:interface-name
```
Comment (Zlatkovsky) [inline types/office-runtime/index.d.ts]: Please remove this tslint comment.  If tslint complains, we should just put that rule in the tslint file

**Pull request merged**